### PR TITLE
465 bug in multi table schema with secondary tables referencing their root table

### LINKED ISSUE #465: Bug in multi-table schema with secondary tables referencing their root table
#### Description
Bug multi-table dans le cas d'un schéma avec référence a la table Root en tant que table externe

Contexte
- cas d'un schéma de type graphe, comme les Molecules, les Atoms et les Bonds référence leur molécule via une une référence
  à la table Molecules qui est Root, ce qui permet de créer les lien entre atomes et liaison et réciproquement
- si on met Molecule en dictionnaire d'analyse, cela se passe bien, et on a le schéma multi-table avec les data paths suivants
  - Molecule (table principale, qui est Root)
  - Atoms
  - Bonds

Problème rencontré
- si on met par exemple Atom en dictionnaire d'analyse, on obtient un schéma "pathologique"
  - Atoms (table principale des atomes)
  - /Molecule
  - /Molecule/Atoms
  - /Molecule/Bonds
- le schéma des molécule est traité comme une table externe, avec sa sous-tables des atomes, déjà utilisée en table principale
  - le fichier des atomes ne peut être utilise deux fois
  - on a toutes la base des molécules chargée en mémoire en tant que table externe
  - ce comportement n'a pas de sens

Pistes de solutions
- en premier lieu, détecter correctement ce cas: celui d'une table référençant sa table parente en tant que table externe
- dans ce cas, couper lien avec la table externe parente, pour permettre l'analyse de la table des atomes seules, sans lien avec les molécules
  - ne présenter que la table Atoms dans le choix des tables d'analyse
  - émettre un warning explicatif
  - ignorer la référence à la table externe lors de l'analyse
- plus tard: proposer une extension des dictionnaires, avec spécification des liens internes d'un flocon

### Context
- Khiops version: V10, V11
- Correction in V11, no need to backport correction to V10

### PR DESCRIPTION
**Pull request by marcboulle**
Review a faire avec celui de la PR https://github.com/KhiopsML/khiops/pull/485

Seuls les 4 derniers commits, propres à la PR, sont à prendre en compte

### CODE REVIEW

File: src/Learning/KWData/KWMTDatabase.cpp [hunk 1]
```diff
@@ -332,40 +345,94 @@ void KWMTDatabase::UpdateMultiTableMappings()
...
 			{
-				// Creation d'un tableau de mapping, pour accuillir les mapping pour la classe externe
-				// en cours de traitement
-				oaCreatedMappings = new ObjectArray;
-				oaAllMainCreatedMappings.Add(oaCreatedMappings);
-
-				// Creation du mapping et memorisation de tous les mapping des sous-classes
+				// Premiere passe de creation du mapping de la table racine externe, avec des containers de travail
+				// Cela permet d'analyser la structure des mappings, sans impacter directement le contenu des containers globaux
+				odWorkingReferenceClasses.RemoveAll();
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.cpp:366]: Perhaps add a comment here, like:
```c++
// On s'assure que les containers soient vides
```
Or perhaps a more precise comment, which states that (unlike before) we make sure the data structures in question are only updated via mutation by `CreateMapping`.

File: src/Learning/KWData/KWMTDatabase.cpp [hunk 2]
```diff
@@ -332,40 +345,94 @@ void KWMTDatabase::UpdateMultiTableMappings()
...
-				mapping = CreateMapping(&odReferenceClasses, &oaRankedReferenceClasses,
-							&odAnalysedCreatedClasses, referenceClass, true,
-							referenceClass->GetName(), &svAttributeName, oaCreatedMappings);
+				mapping = CreateMapping(&odWorkingReferenceClasses, &oaWorkingRankedReferenceClasses,
+							&odWorkingAnalysedCreatedClasses, referenceClass, true,
+							referenceClass->GetName(), &svAttributeName,
+							&oaWorkingCreatedMappings);
 				assert(svAttributeName.GetSize() == 0);
+
+				// On determine si le dictionnaire de reference est utilisablen c'est a dire s'il n'utilise
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.cpp]: s/utilisablen/utilisable/

File: src/Learning/KWData/KWMTDatabase.cpp [hunk 3]
```diff
@@ -332,40 +345,94 @@ void KWMTDatabase::UpdateMultiTableMappings()
...
+				// pas la classe principale dans ses mappings
+				bIsRootDictionaryUsable = true;
+				for (j = 0; j < oaWorkingCreatedMappings.GetSize(); j++)
+				{
+					mapping = cast(KWMTDatabaseMapping*, oaWorkingCreatedMappings.GetAt(j));
+
+					// Transfer des specification de la table mappee si comparaison positive
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.cpp]: s/Transfer/Transfert/

Comment (popescu-v) [reply src/Learning/KWData/KWMTDatabase.cpp]: s/des specification/des specifications/

File: src/Learning/KWData/KWMTDatabase.cpp [hunk 4]
```diff
@@ -332,40 +345,94 @@ void KWMTDatabase::UpdateMultiTableMappings()
...
+
+				// Prise en compte de la classe referencee si elle est utilisable
+				if (bIsRootDictionaryUsable)
+				{
+					// Creation et memorisation d'un tableau de mapping, pour accueillir les mappings
+					// pour la classe externe en cours de traitement
+					oaCreatedMappings = new ObjectArray;
+					oaAllRootCreatedMappings.Add(oaCreatedMappings);
+
+					// Creation du mapping et memorisation de tous les mapping des sous-classes
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.cpp]: tous les mapping**s** (add trailing "s")

File: src/Learning/KWData/KWMTDatabase.cpp [hunk 5]
```diff
@@ -332,40 +345,94 @@ void KWMTDatabase::UpdateMultiTableMappings()
...
+					// Il est plus simple de rappeler la meme methode avec les container globaux
+					// que de fusionner les
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.cpp]: finish the phrase: `// que de fusionner les ....`: something like: `// que de les fusionner avec les containers locaux` (or something more correct / more precise).

File: src/Learning/KWData/KWMTDatabase.cpp [hunk 6]
```diff
@@ -332,40 +345,94 @@ void KWMTDatabase::UpdateMultiTableMappings()
...
+					// Et il n'y a aucn enjeu d'optimisation
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.cpp]: s/aucn/auc**u**n/

File: src/Learning/KWData/KWMTDatabase.h [hunk 7]
```diff
@@ -145,9 +145,10 @@ class KWMTDatabase : public KWDatabase
 	// Creation recursive d'un mapping
 	// Le mapping est chaine avec ses mappings composant.
 	// Le tableaux mapping exhaustifs est egalement egalement mis a jour
-	// Les classes referencees sont memorisees dans un dictionnaire, pour gerer les mappings externes a creer
-	// ulterieurement Les mappings crees recursivement sont memorises dans un tableau
-	// Les classes crees analysees sont egalement memorisees dans un dictionnaire, pour eviter des analyse multiples
+	// Les classes referencees sont memorisees dans un dictionnaire et un tableau,
+	// pour gerer les mappings externes a creer ulterieurement
+	// Les mappings crees recursivement sont memorises dans un tableau
+	// Les classes crees analysees sont egalement memorisees dans un dictionnaire, pour eviter des analyses multiples
```
Comment (popescu-v) [inline src/Learning/KWData/KWMTDatabase.h]: "classes cree**e**s" (add feminine "e").

**Review by popescu-v — CHANGES_REQUESTED** [review]
Some typos and requests for comments.

File: src/Learning/KWData/KWMTDatabase.cpp (hunk 1 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.cpp:366]: Amélioration du commentaire

File: src/Learning/KWData/KWMTDatabase.cpp (hunk 2 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.cpp]: Corrigé

File: src/Learning/KWData/KWMTDatabase.cpp (hunk 3 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.cpp]: Corrigé (deux fois)

File: src/Learning/KWData/KWMTDatabase.cpp (hunk 4 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.cpp]: Corrigé

File: src/Learning/KWData/KWMTDatabase.cpp (hunk 5 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.cpp]: Corrigé

File: src/Learning/KWData/KWMTDatabase.cpp (hunk 6 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.cpp]: Corrigé

File: src/Learning/KWData/KWMTDatabase.h (hunk 7 above)
Comment (marcboulle) [reply src/Learning/KWData/KWMTDatabase.h]: Corrigé

**Review by popescu-v — APPROVED** [review]
LGTM (quick review, mosly commit messages and comments).

**Pull request merged**